src/GraphCoarsen.cpp

```diff
@@ -13,788 +13,1100 @@
+    agg_cdof_vertex.SortIndices();
+
+    return agg_cdof_vertex;
+}
+
+SparseMatrix GraphCoarsen::BuildAggCDofEdge() const
+{
+    int num_aggs = gt_.agg_ext_edge_.Rows();
+    int num_traces = face_cdof_.Cols();
+    int num_cdofs = P_edge_.Cols();
+
+    std::vector<int> indptr(num_aggs + 1);
+    indptr[0] = 0;
+
+    std::vector<int> indices;
+
+    int bubble_counter = 0;
+
+    for (int agg = 0; agg < num_aggs; ++agg)
+    {
+        int num_bubbles_i = vertex_targets_[agg].Cols() - 1;
+        for (int i = 0; i < num_bubbles_i; ++i)
+        {
+            indices.push_back(num_traces + bubble_counter + i);
+        }
+
+        std::vector<int> faces = gt_.agg_face_local_.GetIndices(agg);
+
+        for (auto face : faces)
+        {
+            std::vector<int> face_cdofs = face_cdof_.GetIndices(face);
+
+            for (auto dof : face_cdofs)
             {
-                col = total_num_traces + bubble_counter + j;
-                vertex_target_i.GetColumnReference(j + 1, ref_vec2);
-                entry_value = smoothg::InnerProduct(ref_vec1, ref_vec2);
-
-                if (build_coarse_relation)
-                {
-                    mfem::DenseMatrix& CM_el_loc(CM_el[i]);
-                    CM_el_loc(l, j) = entry_value;
-                    CM_el_loc(j, l) = entry_value;
-                }
-                else
-                {
-                    CoarseM_->Set(row, col, entry_value);
-                    CoarseM_->Set(col, row, entry_value);
-                }
+                indices.push_back(dof);
             }
         }
 
+        indptr[agg + 1] = indices.size();
+
         bubble_counter += num_bubbles_i;
     }
 
-    CoarseD_->Finalize();
+    std::vector<double> data(indices.size(), 1.0);
+
+    return SparseMatrix(std::move(indptr), std::move(indices), std::move(data),
+                        num_aggs, num_cdofs);
+}
+
+ParMatrix GraphCoarsen::BuildEdgeTrueEdge() const
+{
+    int num_faces = face_cdof_.Rows();
+    int num_traces = face_cdof_.Cols();
+    int num_coarse_dofs = P_edge_.Cols();
 
-    if (build_coarse_relation)
-        Agg_cdof_edge_ = make_unique<mfem::SparseMatrix>(Agg_dof_i, Agg_dof_j,
-                                                         Agg_dof_d, nAggs, total_num_traces + bubble_counter);
+    const auto& ftf_diag = gt_.face_true_face_.GetDiag();
 
-    mfem::SparseMatrix face_Agg(smoothg::Transpose(Agg_face));
-    mfem::Array<int> Aggs;
-    mfem::Vector M_v(M_proc_.GetData(), M_proc_.Width()), Mloc_v;
+    int num_true_dofs = num_coarse_dofs - num_traces;
 
-    mfem::Array<int> edge_cdof_marker2, local_Agg_edge_cdof;
-    if (build_coarse_relation)
+    for (int i = 0; i < num_faces; ++i)
     {
-        edge_cdof_marker2.SetSize(Agg_cdof_edge_->Width());
-        edge_cdof_marker2 = -1;
+        if (ftf_diag.RowSize(i) > 0)
+        {
+            num_true_dofs += face_cdof_.RowSize(i);
+        }
     }
```

Comment:
I think `num_true_dofs` is not really used in the rest of the function, and therefore this loop is not needed. Consequently, `ftf_diag` and `num_traces` are also not needed? Or you were using them for other purpose?

Reply:
This was probably just poorly translated from the original master.  I'll remove all the unused parts.